Switch between multiple log views

mkock opened the pull request:
- **PR Description**

* Fixes jesseduffield/lazygit#1363
* Allow switching between up two three log views

Seeing as the last activity related to this issue was over a year ago, I decided to take a stab at this.
The implementation should be fully backwards compatible. Simply add `allBranchesLogCmdAlt1` and/or `allBranchesLogCmdAlt2` to your config file to use them when cycling between log commands using 'a'.
You can even use `allBranchesLogCmdAlt2` together with `allBranchesLogCmd` (skipping `allBranchesLogCmdAlt1`) if you want, it should not affect usability.

This is my first contribution to LazyGit, but I have experience with Go.

Changes:

- Introduced two new optional user config commands, allBranchesLogCmdAlt1+2
- When pressing 'a' in the Status view, cycle between non-empty, non-identical log commands
- There will always be at least one command to run, since allBranhesLogCmd has a default

- **Please check if the PR fulfills these requirements**

* [x] Cheatsheets are up-to-date (run `go generate ./...`)
* [x] Code has been formatted (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#code-formatting))
* [x] Tests have been added/updated (see [here](https://github.com/jesseduffield/lazygit/blob/master/pkg/integration/README.md) for the integration test guide)
* [ ] Text is internationalised (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#internationalisation))
* [ ] Docs (specifically `docs/Config.md`) have been updated if necessary
* [x] You've read through your own file changes for silly mistakes etc

<!--
Be sure to name your PR with an imperative e.g. 'Add worktrees view'
see https://github.com/jesseduffield/lazygit/releases/tag/v0.40.0 for examples
-->

mark2185 commented:
Why constrain ourselves to two? Why not add `allBranchesLogCmds` (plural) instead and make it a `[]string`?

It _is_ backwards compatible, albeit maybe a bit odd to have both singular _and_ plural, but we could easily modify the config to use the updated version.

mkock commented:
> Why constrain ourselves to two? Why not add `allBranchesLogCmds` (plural) instead and make it a `[]string`?
> 
> It _is_ backwards compatible, albeit maybe a bit odd to have both singular _and_ plural, but we could easily modify the config to use the updated version.

It's a good suggestion. And I guess that in the edge case where the user decides to have both `allBranchesLogCmd` _and_ `allBranchesLogCmds` in their config, I could prepend the former to the latter, as a way to honor (rather than ignore) that the user has set a `allBranchesLogCmd`?

mkock commented:
> > Why constrain ourselves to two? Why not add `allBranchesLogCmds` (plural) instead and make it a `[]string`?
> > It _is_ backwards compatible, albeit maybe a bit odd to have both singular _and_ plural, but we could easily modify the config to use the updated version.
> 
> It's a good suggestion. And I guess that in the edge case where the user decides to have both `allBranchesLogCmd` _and_ `allBranchesLogCmds` in their config, I could prepend the former to the latter, as a way to honor (rather than ignore) that the user has set a `allBranchesLogCmd`?

@mark2185 I've updated the PR with your suggestion, please let me know if the change is okay.

mark2185 commented on pkg/integration/tests/status/log_cmd.go:
```diff
@@ -0,0 +1,34 @@
+		t.Views().Status().
+			Focus().
+			Press(keys.Status.AllBranchesLogGraph)
+		t.Views().Main().Content(Contains("view1"))
```
I'd maybe also add the check that it _does not contain view2_ at the same time. Just to be extra sure it's not displaying `view1view2` somehow.

mark2185 commented:
Looks good!

All that remains is updating the docs and cleaning up the history a bit:
- rebasing on `master` would be much cleaner than merging `master` into the branch
- squashing together the 1st and 3rd commit

mkock commented:
> Looks good!
> 
> All that remains is updating the docs and cleaning up the history a bit:
> 
> * rebasing on `master` would be much cleaner than merging `master` into the branch
> * squashing together the 1st and 3rd commit

Apologies, I think I squashed the original commit that you commented on. I've updated the docs and improved the test case. Let me know if there is any other documentation that needs to be updated?

mark2185 commented on docs/Config.md:
```diff
@@ -120,6 +120,7 @@ git:
   fetchAll: true # Pass --all flag when running git fetch. Set to false to fetch only origin (or the current branch's upstream remote if there is one)
   branchLogCmd: 'git log --graph --color=always --abbrev-commit --decorate --date=relative --pretty=medium {{branchName}} --'
   allBranchesLogCmd: 'git log --graph --all --color=always --abbrev-commit --decorate --date=relative  --pretty=medium'
+  allBranchesLogCmds: '' # a list of your favorite log commands (pressing 'a' in the status panel will cycle between them)
```
Shouldn't the default be `[]`?

mark2185 commented:
Other this one minor nitpick, looks good to me! :)

mark2185 commented:
All that's left is appeasing the CI and I think it's good to ship.

mkock commented:
@mark2185 Not sure what's up with the code coverage tool, but everything else is green now

mark2185 commented:
Not sure, to be frank. Also you can squash all the commits now.

@stefanhaller I'm hoping you might have an ace up your sleeve regarding the code coverage.

stefanhaller commented:
It's a known problem, this step currently always fails for PRs created from a fork.

mkock commented:
Done

jesseduffield commented:
Nice work @mkock, thanks for making this

mark2185 commented on pkg/commands/git_commands/branch.go:
```diff
@@ -244,5 +246,17 @@ func (self *BranchCommands) Merge(branchName string, opts MergeOpts) error {
-	return self.cmd.New(str.ToArgv(self.UserConfig.Git.AllBranchesLogCmd)).DontLog()
+	// Only choose between non-empty, non-identical commands
+	candidates := lo.Uniq(lo.WithoutEmpty(append([]string{
+		self.UserConfig.Git.AllBranchesLogCmd,
```
Since `AllBranchesLogCmd` has a default value, this means that if someone sets only the `AllBranchesLogCmds`, it will still prepend the default `AllBranchesLogCmd` to it.

So if someone wants to have only 2 custom log commands, they have to set `AllBranchesLogCmd` to empty as well, which isn't exactly deprecated then, is it?
